### ISSUE TITLE
imread() sometimes leaks file descriptors

### ISSUE THREAD

**Issue by jukkaho** [Itseez/opencv#6162]
With some very specific jpeg images, imread() leaks file descriptors while returning successfully from the call. This was tested with OpenCV 3.1 on Linux.

The following code will demonstrate the problem:
```
#include <opencv2/opencv.hpp>

int main(int argc, char **argv) {
  for (int i = 0; i < 2048; ++i) {
    // Each cv::imread() will leak one file descriptor to bug.jpg
    cv::Mat im = cv::imread("bug.jpg");
    cv::imshow("im", im);
    cv::waitKey(-1);
  }
  return 0;
}
```

Bug is reproducible with an image at http://www.ee.oulu.fi/~jukkaho/opencv-imread-bug/bug.jpg which was created using OpenCV's imwrite() without any special parameters from a decoded video frame. Most of the extracted JPEG files work correctly and do not cause leaks but there are a few that do it every time when decoded by OpenCV.

Leak can be verified by looking at /proc/\<pid\>/fd list of the running process while it's waiting in cv::waitKey():

> ls -la /proc/21856/fd
> total 0
> dr-x------ 2 juki juki  0 Feb 24 19:29 .
> dr-xr-xr-x 8 juki juki  0 Feb 24 19:29 ..
> lrwx------ 1 juki juki 64 Feb 24 19:29 0 -> /dev/pts/17
> lrwx------ 1 juki juki 64 Feb 24 19:29 1 -> /dev/pts/17
> lr-x------ 1 juki juki 64 Feb 24 19:29 10 -> /home/juki/software/tests/fd-leak-test/bug.jpg
> lr-x------ 1 juki juki 64 Feb 24 19:29 11 -> /home/juki/software/tests/fd-leak-test/bug.jpg
> lr-x------ 1 juki juki 64 Feb 24 19:29 12 -> /home/juki/software/tests/fd-leak-test/bug.jpg
> lr-x------ 1 juki juki 64 Feb 24 19:29 13 -> /home/juki/software/tests/fd-leak-test/bug.jpg
> lr-x------ 1 juki juki 64 Feb 24 19:29 14 -> /home/juki/software/tests/fd-leak-test/bug.jpg
> lrwx------ 1 juki juki 64 Feb 24 19:29 2 -> /dev/pts/17
> lrwx------ 1 juki juki 64 Feb 24 19:29 3 -> socket:[39592691]
> lr-x------ 1 juki juki 64 Feb 24 19:29 4 -> /home/juki/software/tests/fd-leak-test/bug.jpg
> lrwx------ 1 juki juki 64 Feb 24 19:29 5 -> anon_inode:[eventfd]
> lrwx------ 1 juki juki 64 Feb 24 19:29 6 -> socket:[39598246]
> lrwx------ 1 juki juki 64 Feb 24 19:29 7 -> socket:[39598252]
> lrwx------ 1 juki juki 64 Feb 24 19:29 8 -> anon_inode:[eventfd]

**Comment by mshabunin** [comment]
Please, check the latest master branch version, it could be fixed by #6121

**Comment by mshabunin** [comment]
Looks like I've found the problem, thanks for the report. Please check the fix made in PR #6166

**Comment by jukkaho** [comment]
Thanks for a quick patch. I can verify that PR #6166 (in addition to #6121) fixes the issue for me.
